File: daemon/daemon_python/window_poller.py
"""
Polls macOS for all on-screen windows via CGWindowListCopyWindowInfo.

Returns a list of window dicts with position, size, owner, layer.
"""

from Quartz import (
    CGWindowListCopyWindowInfo,
    kCGWindowListOptionOnScreenOnly,
    kCGNullWindowID,
    kCGWindowBounds,
    kCGWindowLayer,
    kCGWindowOwnerName,
    kCGWindowOwnerPID,
    kCGWindowName,
    kCGWindowNumber,
)


class WindowPoller:
    """Polls CGWindowListCopyWindowInfo for on-screen windows."""

    def poll(self) -> list[dict]:
        """
        Return all on-screen windows with their metadata.

        Each dict contains:
            owner: str  — process name
            title: str  — window title (may be empty)
            pid: int    — process ID
            layer: int  — window layer (0 = normal, >0 = overlay)
            x, y, width, height: float — window bounds in logical points
            window_id: int — unique window number
        """
        window_list = CGWindowListCopyWindowInfo(
            kCGWindowListOptionOnScreenOnly, kCGNullWindowID
        )

        if not window_list:
            return []

        windows = []
        for w in window_list:
            bounds = w.get(kCGWindowBounds)
            if not bounds:
                continue

            owner = w.get(kCGWindowOwnerName, "")
            if not owner:
                continue

            windows.append({
                "owner": owner,
                "title": w.get(kCGWindowName, "") or "",
                "pid": w.get(kCGWindowOwnerPID, 0),
                "layer": w.get(kCGWindowLayer, 0),
                "x": bounds.get("X", 0),
                "y": bounds.get("Y", 0),
                "width": bounds.get("Width", 0),
                "height": bounds.get("Height", 0),
                "window_id": w.get(kCGWindowNumber, 0),
            })

        return windows

File: daemon/daemon_swift/Sources/PiiMaskDaemon/WindowPoller.swift
/// Polls macOS for all on-screen windows via CGWindowListCopyWindowInfo.
///
/// Direct CoreGraphics call — no PyObjC bridging overhead.
/// Expected: ~2ms avg vs 11.8ms in Python.

import CoreGraphics
import Foundation

struct WindowInfo {
    let owner: String
    let title: String
    let pid: Int32
    let layer: Int32
    let x: Float
    let y: Float
    let width: Float
    let height: Float
    let windowID: UInt32
}

final class WindowPoller {

    func poll() -> [WindowInfo] {
        guard let list = CGWindowListCopyWindowInfo(
            [.optionOnScreenOnly], kCGNullWindowID
        ) as? [[CFString: Any]] else {
            return []
        }

        var windows: [WindowInfo] = []
        windows.reserveCapacity(list.count)

        for dict in list {
            guard let boundsDict = dict[kCGWindowBounds] as? [String: Any],
                  let owner = dict[kCGWindowOwnerName] as? String,
                  !owner.isEmpty
            else { continue }

            let x = (boundsDict["X"] as? NSNumber)?.floatValue ?? 0
            let y = (boundsDict["Y"] as? NSNumber)?.floatValue ?? 0
            let w = (boundsDict["Width"] as? NSNumber)?.floatValue ?? 0
            let h = (boundsDict["Height"] as? NSNumber)?.floatValue ?? 0

            windows.append(WindowInfo(
                owner: owner,
                title: (dict[kCGWindowName] as? String) ?? "",
                pid: (dict[kCGWindowOwnerPID] as? NSNumber)?.int32Value ?? 0,
                layer: (dict[kCGWindowLayer] as? NSNumber)?.int32Value ?? 0,
                x: x, y: y, width: w, height: h,
                windowID: (dict[kCGWindowNumber] as? NSNumber)?.uint32Value ?? 0
            ))
        }

        return windows
    }
}
